rust/datafusion/src/execution/context.rs

```diff
@@ -89,12 +91,18 @@ impl ExecutionConfig {
 }
 
 /// Execution context for registering data sources and executing queries
-pub struct ExecutionContext {
-    datasources: HashMap<String, Box<dyn TableProvider + Send + Sync>>,
-    scalar_functions: HashMap<String, Box<ScalarFunction>>,
+#[derive(Clone)]
+pub struct ExecutionContextState {
+    datasources: Rc<RefCell<HashMap<String, Box<dyn TableProvider + Send + Sync>>>>,
```

Comment:
Can you explain the rationale for using `RefCell` here? It seems to me like `Arc<HashMap<...>>` is easier to reason about. I can't think of an example when running multi-threaded when you would want one thread to borrow mutably -- it seems like an ExecutionContextState would be created/configured by one thread, and then when shared would be read only.

Reply:
So we do need interior mutability here. Users can register new data sources. I can convert to `Arc<Mutex<_>>` instead though.

Reply:
I guess I was thinking that users would register new datasources before/during query planning time (when `ExecutionContextState` was owned by a single thread and therefore the entire `struct` would be mutable, not just a field. I probably misunderstand what is going on